File: coresuite-dotnet-framework-console-vb/generator/GeneratorInvoice.vb
Imports ceTe.DynamicPDF
Imports ceTe.DynamicPDF.PageElements
Imports ceTe.DynamicPDF.PageElements.BarCoding
Imports System
Imports System.Configuration
Imports System.Data.SqlClient

Public Class GeneratorInvoice

    Public Shared order As List(Of GeneratorInvoiceData.Order) = GeneratorInvoiceData.Orders

    Public Shared Function Run(ByVal invoiceNumbers As String()) As Byte()
        'Add the template to the document
        Dim MyDocument As Document = New Document With {
            .Creator = "Invoice",
            .Author = "ceTe Software",
            .Title = "Invoice",
            .Template = MyInvoice.GetTemplate
        }

        ' Add Invoices to the document
        Dim item As String
        Dim data As GeneratorInvoiceData.Order

        For Each item In invoiceNumbers
            For Each data In order
                If item = data.OrderID.ToString Then
                    Dim invoice As MyInvoice = New MyInvoice()
                    invoice.Draw(data, MyDocument)
                End If
            Next
        Next

        'Outputs the Invoices to the current web MyPage
        Try
            Return MyDocument.Draw()
        Catch empty As EmptyDocumentException
            'Show an empty MyDocument error
            Return ShowErrorDocument("Document is empty. No invoices were selected.", "")
        Catch general As Exception
            'Show a general error
            Return ShowErrorDocument(general.Message, general.StackTrace)
        End Try
    End Function

    Private Shared Function ShowErrorDocument(ByVal message As String, ByVal stackTrace As String) As Byte()
        'Shows the error in a PDF MyDocument
        Dim MyDocument As Document = New Document
        Dim MyPage As Page = New Page(PageSize.Letter, PageOrientation.Portrait)

        Dim MyMessage As TextArea = New TextArea(message, 0, 0, 512, 36, Font.Helvetica, 18, RgbColor.Red)
        MyMessage.Height = MyMessage.GetRequiredHeight()
        MyPage.Elements.Add(MyMessage)

        If stackTrace.Trim().Length > 0 Then
            Dim StackTraceTop As Single = MyMessage.Y + MyMessage.Height + 20

            Dim StackTraceMessage As TextArea = New TextArea(stackTrace, 0, StackTraceTop, 512, 692 - StackTraceTop, Font.Helvetica, 10)

            MyPage.Elements.Add(New ceTe.DynamicPDF.PageElements.Label("StackTrace:", 0, StackTraceTop - 12, 512, 12, Font.HelveticaBold, 10))
            MyPage.Elements.Add(StackTraceMessage)
        End If
        MyDocument.Pages.Add(MyPage)
        Return MyDocument.Draw()
    End Function
End Class

Public Class MyInvoice

    Private subTotal As Decimal = 0
    Private freight As Decimal = 0
    Private Shared MyTemplate As Template = New Template
    Private Shared pageTemplateImagesSet As Boolean = False
    Private Shared backgroundColor As WebColor = New WebColor("#E0E0FF")
    Private Shared totalBGColor As WebColor = New WebColor("#FFC0C0")
    Private Shared thankYouText As WebColor = New WebColor("#000080")

    Shared Sub New()
        ' Uncomment the line below to add a layout grid to the pages
        'template.Elements.Add( New LayoutGrid() );

        ' Top part of Invoice
        MyTemplate.Elements.Add(New Label("Northwind Traders\n1234 International Drive\nAnywhere, Earth  ABC123", 56, 0, 200, 44, Font.Helvetica))
        MyTemplate.Elements.Add(New Label("Invoice", 0, 0, 540, 18, Font.HelveticaBold, 18, TextAlign.Right))
        MyTemplate.Elements.Add(New PageNumberingLabel("Page %%SP%% of %%ST%% ", 450, 253, 90, 20, Font.HelveticaBold, 12, TextAlign.Center))

        ' Add Invoice Details Template
        MyTemplate.Elements.Add(GetDetailsGroup())

        ' Add BillTo Template
        MyTemplate.Elements.Add(GetBillToGroup())

        ' Add ShipTo Template
        MyTemplate.Elements.Add(GetShipToGroup())

        ' Add Line Item Template
        MyTemplate.Elements.Add(GetLineItemGroup())
    End Sub

    Public Sub New()
        SetPageTemplateImage()
    End Sub


    Public Shared Function GetTemplate() As Template
        Return MyTemplate
    End Function

    Private Shared Function GetDetailsGroup() As Group
        'Returns a group containing the details template
        Dim MyGroup As Group = New Group

        MyGroup.Add(New Rectangle(340, 24, 200, 72, Grayscale.Black, backgroundColor, 0.5F))
        MyGroup.Add(New Label("Order ID:", 343, 25, 85, 12, Font.HelveticaBold, 12, TextAlign.Right))
        MyGroup.Add(New Label("Order Date:", 343, 39, 85, 12, Font.HelveticaBold, 12, TextAlign.Right))
        MyGroup.Add(New Label("Customer ID:", 343, 53, 85, 12, Font.HelveticaBold, 12, TextAlign.Right))
        MyGroup.Add(New Label("Shipped Date:", 343, 67, 85, 12, Font.HelveticaBold, 12, TextAlign.Right))
        MyGroup.Add(New Label("Shipped Via:", 343, 81, 85, 12, Font.HelveticaBold, 12, TextAlign.Right))

        Return MyGroup
    End Function

    Private Shared Function GetBillToGroup() As Group
        'Returns a group containing the bill to template
        Dim MyGroup As Group = New Group

        MyGroup.Add(New Rectangle(25, 120, 200, 90, 0.5F))
        MyGroup.Add(New Line(25, 136, 225, 136, 0.5F))
        MyGroup.Add(New Label("Bill To:", 28, 121, 200, 12, Font.HelveticaBold, 12))

        Return MyGroup
    End Function

    Private Shared Function GetShipToGroup() As Group
        'Returns a group containing the ship to template
        Dim MyGroup As Group = New Group

        MyGroup.Add(New Rectangle(315, 120, 200, 90, 0.5F))
        MyGroup.Add(New Line(315, 136, 515, 136, 0.5F))
        MyGroup.Add(New Label("Ship To:", 318, 121, 200, 12, Font.HelveticaBold, 12))

        Return MyGroup
    End Function

    Private Shared Function GetLineItemGroup() As Group
        'Returns a group containing the line items template
        Dim MyGroup As Group = New Group

        Dim I As Integer
        For I = 0 To 10 - 1 Step I + 1
            MyGroup.Add(New Rectangle(0, 306 + I * 36, 540, 18, Grayscale.Black, backgroundColor, 0.0F))
        Next

        MyGroup.Add(New Rectangle(450, 250, 90, 20, 0.5F))
        MyGroup.Add(New Rectangle(450, 702, 90, 18, Grayscale.Black, totalBGColor, 0.0F))
        MyGroup.Add(New Rectangle(0, 270, 540, 450, 0.5F))
        MyGroup.Add(New Line(0, 288, 540, 288, 0.5F))
        MyGroup.Add(New Line(0, 666, 540, 666, 0.5F))
        MyGroup.Add(New Line(60, 270, 60, 666, 0.5F))
        MyGroup.Add(New Line(360, 270, 360, 720, 0.5F))
        MyGroup.Add(New Line(450, 270, 450, 720, 0.5F))
        MyGroup.Add(New Line(450, 702, 540, 702, 0.5F))
        MyGroup.Add(New Label("Quantity", 0, 272, 60, 12, Font.HelveticaBold, 12, TextAlign.Center))
        MyGroup.Add(New Label("Description", 60, 272, 300, 12, Font.HelveticaBold, 12, TextAlign.Center))
        MyGroup.Add(New Label("Unit Price", 360, 272, 90, 12, Font.HelveticaBold, 12, TextAlign.Center))
        MyGroup.Add(New Label("Price", 450, 272, 90, 12, Font.HelveticaBold, 12, TextAlign.Center))

        MyGroup.Add(New Label("Thank you for your purchase." & vbCrLf & "We appreciate your business.", 5, 672, 350, 54, Font.HelveticaBold, 18, thankYouText))
        MyGroup.Add(New Label("Sub Total", 364, 668, 82, 12, Font.HelveticaBold, 12, TextAlign.Right))
        MyGroup.Add(New Label("Freight", 364, 686, 82, 12, Font.HelveticaBold, 12, TextAlign.Right))
        MyGroup.Add(New Label("Total", 364, 704, 82, 12, Font.HelveticaBold, 12, TextAlign.Right))

        Return MyGroup
    End Function

    Private Sub SetPageTemplateImage()
        ' Adds the image to page template if it Is Not already added
        If Not pageTemplateImagesSet Then

            MyTemplate.Elements.Add(New Image(Util.GetResourcePath("Images/DPDFLogo.png"), 0, 0, 0.16F))
            pageTemplateImagesSet = True
        End If
    End Sub

    Public Sub Draw(ByVal Data As GeneratorInvoiceData.Order, ByVal MyDocument As Document)
        'Adds the invoice if there is Data
        If Not Data Is Nothing Then
            'Sets the Freight amount
            freight = CType(Data.Freight, Decimal)

            'Draws the invoice data, returns a MyPage object if it is the last MyPage
            Dim i As Integer = 0
            Dim LastPage As Page = DrawInvoiceData(Data, MyDocument, i)
            'Draws aditional MyPages if necessary
            While LastPage Is Nothing
                LastPage = DrawInvoiceData(Data, MyDocument, i)
            End While

            'Draws the totals to the bottom of the last MyPage of the Invoice
            DrawTotals(Data, MyDocument.Pages.Item(MyDocument.Pages.Count - 1))
        End If
    End Sub

    Private Function DrawInvoiceData(ByVal Data As GeneratorInvoiceData.Order, ByVal MyDocument As Document, ByRef start As Integer) As Page
        'Tracks if the invoice is finished
        Dim InvoiceFinished As Boolean = True
        'Tracks the y position on the MyPage
        Dim YOffset As Single = 288

        'Create a MyPage for the Invoice
        Dim MyPage As Page = New Page(PageSize.Letter, PageOrientation.Portrait, 36.0F)
        'Add Details to the Invoice
        DrawInvoiceDetails(Data, MyPage)
        'Add bill to address
        MyPage.Elements.Add(New TextArea(Data.BillTo, 28, 138, 194, 70, Font.Helvetica, 12))
        'Add ship to address
        MyPage.Elements.Add(New TextArea(Data.ShipTo, 318, 138, 194, 70, Font.Helvetica, 12))

        'Add line items to the Invoice
        For i As Integer = start To Data.OrderDetails.Count - 1
            DrawLineItem(Data.OrderDetails.Item(i), MyPage, YOffset)
            'Break if at the bottom of the MyPage
            If YOffset >= 666 Then
                InvoiceFinished = False
                start = i + 1
                Exit For
            End If
        Next i

        'Add the MyPage to the MyDocument
        MyDocument.Pages.Add(MyPage)

        'If Invoice is finished return the MyPage else return null so another MyPage will be added
        If InvoiceFinished Then
            Return MyPage
        Else
            MyPage.Elements.Add(New Label("Continued...", 454, 704, 82, 12, Font.Helvetica, 12, TextAlign.Right))
            Return Nothing
        End If
    End Function

    Private Sub DrawInvoiceDetails(ByVal Data As GeneratorInvoiceData.Order, ByVal MyPage As Page)
        'Adds Invoice details to the MyPage
        MyPage.Elements.Add(New Label(Data.OrderID.ToString(), 437, 25, 100, 12, Font.Helvetica, 12))
        MyPage.Elements.Add(New Label((CType(Data.OrderDate, DateTime)).ToString("d MMM yyyy"), 437, 39, 100, 12, Font.Helvetica, 12))
        MyPage.Elements.Add(New Label(Data.CustomerID.ToString(), 437, 53, 100, 12, Font.Helvetica, 12))
        'Dim ShippedDate As Label = Nothing
        If Not Data.ShippedDate = Nothing Then
            MyPage.Elements.Add(New Label((CType(Data.ShippedDate, DateTime)).ToString("d MMM yyyy"), 437, 67, 100, 12, Font.Helvetica, 12))
        End If
        MyPage.Elements.Add(New Label(Data.ShipperName.ToString(), 437, 81, 100, 12, Font.Helvetica, 12))
        MyPage.Elements.Add(New Interleaved25(Data.OrderID.ToString(), 380, 4, 18, False))
    End Sub

    Private Sub DrawLineItem(ByVal Detail As GeneratorInvoiceData.OrderDetail, ByVal MyPage As Page, ByRef YOffset As Single)
        'Adds a line item to the invoice
        Dim Quantity As Int16 = CType(Detail.Quantity, Int16)
        Dim UnitPrice As Decimal = CType(Detail.UnitPrice, Decimal)
        Dim LineTotal As Decimal = UnitPrice * Quantity
        subTotal += LineTotal

        MyPage.Elements.Add(New Label(Quantity.ToString(), 4, 3 + YOffset, 52, 12, Font.Helvetica, 12, TextAlign.Right))
        MyPage.Elements.Add(New Label(Detail.ProductName.ToString(), 64, 3 + YOffset, 292, 12, Font.Helvetica, 12))
        MyPage.Elements.Add(New Label(UnitPrice.ToString("#,##0.00"), 364, 3 + YOffset, 82, 12, Font.Helvetica, 12, TextAlign.Right))
        MyPage.Elements.Add(New Label(LineTotal.ToString("#,##0.00"), 454, 3 + YOffset, 82, 12, Font.Helvetica, 12, TextAlign.Right))

        YOffset += 18
    End Sub

    Private Sub DrawTotals(ByVal Data As GeneratorInvoiceData.Order, ByVal MyPage As Page)
        'Add totals to the bottom of the Invoice
        Dim GrandTotal As Decimal = subTotal + freight
        MyPage.Elements.Add(New Label(subTotal.ToString("#,##0.00"), 454, 668, 82, 12, Font.Helvetica, 12, TextAlign.Right))
        MyPage.Elements.Add(New Label(freight.ToString("#,##0.00"), 454, 686, 82, 12, Font.Helvetica, 12, TextAlign.Right))
        MyPage.Elements.Add(New Label(GrandTotal.ToString("#,##0.00"), 454, 704, 82, 12, Font.Helvetica, 12, TextAlign.Right))
    End Sub
End Class

File: coresuite-dotnet-core-console-cs/generator/GeneratorInvoice.cs
using ceTe.DynamicPDF;
using ceTe.DynamicPDF.PageElements;
using ceTe.DynamicPDF.PageElements.BarCoding;
using System;
using System.Collections.Generic;
using System.Data.SqlClient;

namespace coresuite_dotnet_core_console_cs.generator
{
    public class GeneratorInvoice
    {
        public static List<GeneratorInvoiceData.Order> order = GeneratorInvoiceData.Orders;
      
        public static byte[] Run(string[] invoiceNumbers)
        {
            Document document = new Document
            {
                Creator = "Invoice",
                Author = "ceTe Software",
                Title = "Invoice",

                //Add the template to the document
                Template = MyInvoice.GetTemplate
            };

            // Add Invoices to the document           
            foreach (string item in invoiceNumbers)
            {
                foreach (GeneratorInvoiceData.Order data in order)
                {                              
                        if (item == data.OrderID.ToString())
                        {
                            MyInvoice invoice = new MyInvoice();
                            invoice.Draw(data, document);
                        }
                }
            }

            // Outputs the Invoices to the current web page
            try
            {
                return document.Draw();
            }
            catch (EmptyDocumentException)
            {
                // Show an empty document error
                return ShowErrorDocument("Document is empty. No invoices were selected.", "");
            }
            catch (Exception excGeneral)
            {
                // Show a general error
                return ShowErrorDocument(excGeneral.Message, excGeneral.StackTrace);
            }
        }

        private static byte[] ShowErrorDocument(string message, string stackTrace)
        {
            // Shows the error in a PDF document
            Document document = new Document();
            Page page = new Page(PageSize.Letter, PageOrientation.Portrait);

            TextArea messageArea = new TextArea(message, 0, 0, 512, 36, Font.Helvetica, 18, RgbColor.Red);
            messageArea.Height = messageArea.GetRequiredHeight();
            page.Elements.Add(messageArea);

            if (stackTrace.Trim().Length > 0)
            {
                float stackTraceTop = messageArea.Y + messageArea.Height + 20;

                TextArea stackTraceArea = new TextArea(stackTrace, 0, stackTraceTop, 512, 692 - stackTraceTop, Font.Helvetica, 10);

                page.Elements.Add(new Label("StackTrace:", 0, stackTraceTop - 12, 512, 12, Font.HelveticaBold, 10));
                page.Elements.Add(stackTraceArea);
            }
            document.Pages.Add(page);
            return document.Draw();
        }
    }


        public class MyInvoice
        {
            #region Member Variables

            private decimal subTotal = 0;
            private decimal freight = 0;
            private static Template template = new Template();
            private static bool pageTemplateImagesSet = false;
            private static WebColor backgroundColor = new WebColor("#E0E0FF");
            private static WebColor totalBGColor = new WebColor("#FFC0C0");
            private static WebColor thankYouText = new WebColor("#000080");
            public static List<GeneratorInvoiceData.Order> order = GeneratorInvoiceData.Orders;
            #endregion

            #region Constructors

            static MyInvoice()
            {
                // Uncomment the line below to add a layout grid to the pages
                //template.Elements.Add( new LayoutGrid() );

                // Top part of Invoice
                template.Elements.Add(new Label("Northwind Traders\n1234 International Drive\nAnywhere, Earth  ABC123", 56, 0, 200, 44, Font.Helvetica));
                template.Elements.Add(new Label("Invoice", 0, 0, 540, 18, Font.HelveticaBold, 18, TextAlign.Right));
                template.Elements.Add(new PageNumberingLabel("Page %%SP%% of %%ST%% ", 450, 253, 90, 20, Font.HelveticaBold, 12, TextAlign.Center));

                // Add Invoice Details Template
                template.Elements.Add(GetDetailsGroup());

                // Add BillTo Template
                template.Elements.Add(GetBillToGroup());

                // Add ShipTo Template
                template.Elements.Add(GetShipToGroup());

                // Add Line Item Template
                template.Elements.Add(GetLineItemGroup());
            }

            public MyInvoice()
            {
                SetPageTemplateImage();
            }

            #endregion

            public static Template GetTemplate
            {
                get { return template; }
            }

            #region Static Methods

            private static Group GetDetailsGroup()
            {
                // Returns a group containing the details template
                Group group = new Group();
                group.Add(new Rectangle(340, 24, 200, 72, RgbColor.Black, backgroundColor, 0.5f));
                group.Add(new Label("Order ID:", 343, 25, 85, 12, Font.HelveticaBold, 12, TextAlign.Right));
                group.Add(new Label("Order Date:", 343, 39, 85, 12, Font.HelveticaBold, 12, TextAlign.Right));
                group.Add(new Label("Customer ID:", 343, 53, 85, 12, Font.HelveticaBold, 12, TextAlign.Right));
                group.Add(new Label("Shipped Date:", 343, 67, 85, 12, Font.HelveticaBold, 12, TextAlign.Right));
                group.Add(new Label("Shipped Via:", 343, 81, 85, 12, Font.HelveticaBold, 12, TextAlign.Right));
                return group;
            }

            private static Group GetBillToGroup()
            {
                // Returns a group containing the bill to template
                Group group = new Group();
                group.Add(new Rectangle(25, 120, 200, 90, 0.5f));
                group.Add(new Line(25, 136, 225, 136, 0.5f));
                group.Add(new Label("Bill To:", 28, 121, 200, 12, Font.HelveticaBold, 12));
                return group;
            }

            private static Group GetShipToGroup()
            {
                // Returns a group containing the ship to template
                Group group = new Group();
                group.Add(new Rectangle(315, 120, 200, 90, 0.5f));
                group.Add(new Line(315, 136, 515, 136, 0.5f));
                group.Add(new Label("Ship To:", 318, 121, 200, 12, Font.HelveticaBold, 12));
                return group;
            }

            private static Group GetLineItemGroup()
            {
                // Returns a group containing the line items template
                Group group = new Group();

                for (int i = 0; i < 10; i++)
                {
                    group.Add(new Rectangle(0, 306 + i * 36, 540, 18, RgbColor.Black, backgroundColor, 0.0F));
                }

                group.Add(new Rectangle(450, 250, 90, 20, 0.5f));
                group.Add(new Rectangle(450, 702, 90, 18, RgbColor.Black, totalBGColor, 0.0F));
                group.Add(new Rectangle(0, 270, 540, 450, 0.5f));
                group.Add(new Line(0, 288, 540, 288, 0.5f));
                group.Add(new Line(0, 666, 540, 666, 0.5f));
                group.Add(new Line(60, 270, 60, 666, 0.5f));
                group.Add(new Line(360, 270, 360, 720, 0.5f));
                group.Add(new Line(450, 270, 450, 720, 0.5f));
                group.Add(new Line(450, 702, 540, 702, 0.5f));
                group.Add(new Label("Quantity", 0, 272, 60, 12, Font.HelveticaBold, 12, TextAlign.Center));
                group.Add(new Label("Description", 60, 272, 300, 12, Font.HelveticaBold, 12, TextAlign.Center));
                group.Add(new Label("Unit Price", 360, 272, 90, 12, Font.HelveticaBold, 12, TextAlign.Center));
                group.Add(new Label("Price", 450, 272, 90, 12, Font.HelveticaBold, 12, TextAlign.Center));
                group.Add(new Label("Thank you for your purchase.\nWe appreciate your business.", 5, 672, 350, 54, Font.HelveticaBold, 18, thankYouText));
                group.Add(new Label("Sub Total", 364, 668, 82, 12, Font.HelveticaBold, 12, TextAlign.Right));
                group.Add(new Label("Freight", 364, 686, 82, 12, Font.HelveticaBold, 12, TextAlign.Right));
                group.Add(new Label("Total", 364, 704, 82, 12, Font.HelveticaBold, 12, TextAlign.Right));

                return group;
            }

            #endregion

            #region Private Methods

            private void SetPageTemplateImage()
            {
                // Adds the image to page template if it is not already added
                if (!pageTemplateImagesSet)
                {
                    template.Elements.Add(new Image(Util.GetResourcePath("Images/DPDFLogo.png"), 0, 0, 0.16f));
                    pageTemplateImagesSet = true;
                }
            }

            public void Draw(GeneratorInvoiceData.Order data, Document document)
            {
            //Adds the invoice if there is data
            if (data != null)
                {
                    // Sets the Freight amount
                    freight = (decimal)data.Freight;

                    // Draws the invoice data, returns a page object if it is the last page
                    int i = 0;
                    Page lastPage = DrawInvoiceData(data, document, ref i);

                    // Draws aditional pages if necessary
                    while (lastPage == null)
                    {
                        lastPage = DrawInvoiceData(data, document, ref i);
                    }
                    // Draws the totals to the bottom of the last page of the Invoice
                    DrawTotals(data, document.Pages[document.Pages.Count - 1]);
                }
            }

            private Page DrawInvoiceData(GeneratorInvoiceData.Order data, Document document, ref int start)
            {
                // Tracks if the invoice is finished
                bool invoiceFinished = true;

                // Tracks the y position on the page
                float yOffset = 288;

                // Create a page for the Invoice
                Page page = new Page(PageSize.Letter, PageOrientation.Portrait, 36.0f);

                // Add Details to the Invoice
                DrawInvoiceDetails(data, page);

                // Add bill to address         
                page.Elements.Add(new TextArea(data.BillTo, 28, 139, 194, 70, Font.Helvetica, 12));

                // Add ship to address           
                page.Elements.Add(new TextArea(data.ShipTo, 318, 139, 194, 70, Font.Helvetica, 12));

                // Add line items to the Invoice               
                for (int i = start; i < data.OrderDetails.Count; i++)
                {
                    DrawLineItem(data.OrderDetails[i], page, ref yOffset);
                    // Break if at the bottom of the page
                    if (yOffset >= 666)
                    {
                        invoiceFinished = false;
                        start = i + 1;
                        break;
                    }
                }
            
            // Add the page to the document
            document.Pages.Add(page);

                // If Invoice is finished return the page else return null so another page will be added
                if (invoiceFinished)
                {
                    return page;
                }
                else
                {
                    page.Elements.Add(new Label("Continued...", 454, 704, 82, 12, Font.Helvetica, 12, TextAlign.Right));
                    return null;
                }
            }

            private void DrawInvoiceDetails(GeneratorInvoiceData.Order data, Page page)
            {
                // Adds Invoice details to the page
                page.Elements.Add(new Label(data.OrderID.ToString(), 437, 25, 100, 12, Font.Helvetica, 12));
                page.Elements.Add(new Label(((DateTime)data.OrderDate).ToString("d MMM yyyy"), 437, 39, 100, 12, Font.Helvetica, 12));
                page.Elements.Add(new Label(data.CustomerID.ToString(), 437, 53, 100, 12, Font.Helvetica, 12));

                if (!(data.ShippedDate == null))
                {
                    page.Elements.Add(new Label(((DateTime)data.ShippedDate).ToString("d MMM yyyy"), 437, 67, 100, 12, Font.Helvetica, 12));
                }

                page.Elements.Add(new Label(data.ShipperName.ToString(), 437, 81, 100, 12, Font.Helvetica, 12));
                page.Elements.Add(new Interleaved25(data.OrderID.ToString(), 380, 4, 18, false));
            }

            private void DrawLineItem(GeneratorInvoiceData.OrderDetail detail, Page page, ref float yOffset)
            {
                // Adds a line item to the invoice
                Int16 quantity = (Int16)detail.Quantity;
                decimal unitPrice = (decimal)detail.UnitPrice;
                decimal lineTotal = unitPrice * quantity;
                subTotal += lineTotal;

                page.Elements.Add(new Label(quantity.ToString(), 4, 3 + yOffset, 52, 12, Font.Helvetica, 12, TextAlign.Right));
                page.Elements.Add(new Label(detail.ProductName.ToString(), 64, 3 + yOffset, 292, 12, Font.Helvetica, 12));
                page.Elements.Add(new Label(unitPrice.ToString("#,##0.00"), 364, 3 + yOffset, 82, 12, Font.Helvetica, 12, TextAlign.Right));
                page.Elements.Add(new Label(lineTotal.ToString("#,##0.00"), 454, 3 + yOffset, 82, 12, Font.Helvetica, 12, TextAlign.Right));
                yOffset += 18;
            }

            private void DrawTotals(GeneratorInvoiceData.Order data, Page page)
            {
                // Add totals to the bottom of the Invoice
                decimal grandTotal = subTotal + freight;
                page.Elements.Add(new Label(subTotal.ToString("#,##0.00"), 454, 668, 82, 12, Font.Helvetica, 12, TextAlign.Right));
                page.Elements.Add(new Label(freight.ToString("#,##0.00"), 454, 686, 82, 12, Font.Helvetica, 12, TextAlign.Right));
                page.Elements.Add(new Label(grandTotal.ToString("#,##0.00"), 454, 704, 82, 12, Font.Helvetica, 12, TextAlign.Right));
            }
            #endregion
        }
    }


